Support azure signtool

marcoesters opened the pull request:
### Description

Add support for AzureSignTool to sign Windows installers. `AzureSignTool` uses Azure key vaults, which do not use certificate files on disk. To achieve this, this PR:

* Adds a `windows_signing_tool` keyword that selects the tool to sign. This key determines whether to sign the installer instead of `signing_certificate`, which makes signing more extensible.
* Creates classes for Windows signing tools.
* Adds documentation for all environment variables for both `SignTool` and `AzureSignTool`.
* Allows `signtool.exe` to use a different digest algorithms.

Since this requires access to Azure, tests cannot be added for this feature.

Closes #767.


### Checklist - did you ...

- [X] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
- [X] Add / update necessary tests?
- [X] Add / update outdated documentation?

jaimergp commented on docs/source/howto.md:
```diff
@@ -38,21 +38,63 @@ On Windows, you can also add extra pages to the installer. This is an advanced o
+:::{note}
+
+If neither `AZURE_SIGNTOOL_KEY_VAULT_ACCESSTOKEN` nor `AZURE_SIGNTOOL_KEY_VAULT_SECRET` are set, `constructor` will use a Managed Identity (`-kvm` CLI option).
+
```
```suggestion
:::
```

jaimergp commented on docs/source/howto.md:
```diff
@@ -38,21 +38,63 @@ On Windows, you can also add extra pages to the installer. This is an advanced o
+
+If neither `AZURE_SIGNTOOL_KEY_VAULT_ACCESSTOKEN` nor `AZURE_SIGNTOOL_KEY_VAULT_SECRET` are set, `constructor` will use a Managed Identity (`-kvm` CLI option).
+
+### Signing and notarizing pkg installers
```
```suggestion
### Signing and notarizing PKG installers
```

jaimergp commented on news/771-add-azure-signtool-support:
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* Add support for AzureSignTool to sign Windows installers.
```
```suggestion
* Add support for AzureSignTool to sign Windows installers. (#767 via #771)
```

jaimergp commented on constructor/construct.py:
```diff
@@ -792,6 +801,15 @@ def verify(info):
                          f"must be an instance of {types_str}")
+    if signtool := info.get("windows_signing_tool"):
+        if signtool not in WIN_SIGNTOOLS:
+            sys.exit(
```
This reminds me we need to start raising exceptions instead of sys.exits, but that's for another PR.

jaimergp commented on constructor/main.py:
```diff
@@ -109,6 +109,12 @@ def main_build(dir_path, output_dir='.', platform=cc_platform,
 
+    # Normalize name and set default value
+    if info.get("windows_signing_tool"):
+        info["windows_signing_tool"] = info["windows_signing_tool"].lower().replace(".exe", "")
```
At the same time, if we are validating the input in construct.py:805, why do we need normalization here? 🤔 For library usage? In that case it should be the same logic (or move the logic here).

marcoesters replied:
> What happens if a user tries to set this to a full path? 

`constructor` will and should fail. `windows_signing_tool` is the name. Otherwise, I would have to infer from the binary what the appropriate signing too is.

>  In that case it should be the same logic

Yes, that's true, good catch!

> (or move the logic here).

I'm not sure about this. Setting default values and manipulating input is done in `main_build`. `validate()` performs validation here. If we move this line into `validate()`, we are mixing scopes, which makes things harder to track.

jaimergp commented on constructor/main.py:
```diff
@@ -109,6 +109,12 @@ def main_build(dir_path, output_dir='.', platform=cc_platform,
+    if info.get("windows_signing_tool"):
+        info["windows_signing_tool"] = info["windows_signing_tool"].lower().replace(".exe", "")
+    elif info.get("signing_certificate"):
+        info["windows_signing_tool"] = "signtool"
```
We should check if we are overriding something here, and warn appropriately.

marcoesters replied:
That check is done in line 113 - if `windows_signing_tool` is set in any way, we won't get to `elif`.

jaimergp commented on constructor/signing.py:
```diff
@@ -0,0 +1,222 @@
+        return command
+
+    def verify_signing_tool(self):
+        self._verify_tool_is_available()
```
```suggestion
        super()._verify_tool_is_available()
```

jaimergp commented on constructor/signing.py:
```diff
@@ -0,0 +1,222 @@
+        super().__init__(os.environ.get("CONSTRUCTOR_SIGNTOOL_PATH", "AzureSignTool"))
+
+    def get_signing_command(self) -> str:
+        def _check_required_env_vars(env_vars):
+            missing_vars = {var for var in env_vars if var not in os.environ}
+            if missing_vars:
+                raise RuntimeError(
+                    f"Missing required environment variables {', '.join(missing_vars)}."
+                )
```
I think this can be easily moved to the base class as its own (static?) method.

marcoesters replied:
Since this is a very general helper function, I moved it to utils

jaimergp commented on constructor/signing.py:
```diff
@@ -0,0 +1,222 @@
+            f" -td {timestamp_digest}"
+            f" -fd {file_digest}"
+        )
+        # There are three ways to sign:
+        #   1. Access token
+        #   2. Secret (requires tenant ID)
+        #   3. Managed identity (requires prior login to Azure)
```
Can we `log.info` which method was detected?

jaimergp commented on constructor/signing.py:
```diff
@@ -0,0 +1,222 @@
+
+    def verify_signature(self, installer_file: Union[str, Path]):
+        """
+   `    https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.signaturestatus
```
```suggestion
        https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.signaturestatus
```

jaimergp commented on constructor/signing.py:
```diff
@@ -0,0 +1,222 @@
+        # but stderr will be non-empty on errors
+        if proc.stderr:
+            logger.error(proc.stderr)
+            return
```
Should we raise here? 🤔

jaimergp commented on constructor/winexe.py:
```diff
@@ -217,27 +210,7 @@ def uninstall_menus_commands(info):
-
-
-def make_nsi(info, dir_path, extra_files=None, temp_extra_files=None):
+def make_nsi(info, dir_path, extra_files=None, temp_extra_files=None, signing_tool=None):
```
Can we add some typing? There are a few things called `signtool` or similar now and it can get confusing. I think this expects an instance of the new class in this PR?

jaimergp reviewed (commented):
Very nice work. I'm a little concerned about the lack of tests, so I wonder if it's feasible to have a some secrets added to a dummy vault (or something that implements the same API). However, that might be too complicated and this feature will be used routinely by Anaconda anyway, so I guess we'll know when it breaks.

Double check that the docs render correctly locally, I found a little typo with the admonition syntax. Other than that, this is good to go after addressing the comments.

marcoesters commented:
I added integration tests and secrets were added to the repository. However, the secrets will not be available for anybody who submits a PR from the fork, which will cause the tests to fail. The tests do execute the correct command though: https://github.com/conda/constructor/actions/runs/8760110611/job/24044501617#step:9:2133

Those tests will have to be skipped until merged into `main`. So, we will have to add 
As discussed with @jaimergp offline, the best way to go here is to add test failure reporting as they do for `conda-libmambda-solver`: https://github.com/conda/conda-libmamba-solver/blob/efa3b84141e4ff777928aaae952fac417b059091/.github/workflows/tests.yml#L680-L690

This requires a token, so I will convert this PR into draft until that infrastructure is in place.

Additional things I found:

* The environment is not copied into the subprocess on Windows when no extra environment variables are set. On POSIX, this is handled by the subprocess module, but not for Windows. So, I copy the environment explicitly for all to get conistent behavior.
* The debugging section was not in any TOC tree for the documentations, so I added it.

jaimergp commented on .github/workflows/main.yml:
```diff
@@ -94,6 +94,10 @@ jobs:
           environment-file: dev/environment.yml
           python-version: ${{ matrix.python-version }}
+      - name: Install AzureSignTool
+        if: ${{ matrix.os == 'windows' }}
```
```suggestion
        if: matrix.os == 'windows'
```
These can be omitted in `if` clauses, I think.

jaimergp commented on constructor/signing.py:
```diff
@@ -0,0 +1,220 @@
+        )
+        command = (
+            f"{win_str_esc(self.executable)} sign /f {win_str_esc(self.certificate_file)} "
+            f"/tr {win_str_esc(timestamp_server)} /td {timestamp_digest} /fd {file_digest}"
```
Just noting here that no quotes are needed because we don't expect spaces here.

jaimergp commented on constructor/signing.py:
```diff
@@ -0,0 +1,220 @@
+        command = (
+            f"{self.executable} sign -v"
+            ' -kvu %AZURE_SIGNTOOL_KEY_VAULT_URL%'
+            ' -kvc %AZURE_SIGNTOOL_KEY_VAULT_CERTIFICATE%'
```
Should we quote this?

marcoesters replied:
We can. I just tested it locally and the test worked fine with quotes, too.

jaimergp commented on constructor/signing.py:
```diff
@@ -0,0 +1,220 @@
+            )
+            check_required_env_vars(required_env_vars)
+            command += (
+                " -kvi %AZURE_SIGNTOOL_KEY_VAULT_CLIENT_ID%"
+                " -kvt %AZURE_SIGNTOOL_KEY_VAULT_TENANT_ID%"
+                " -kvs %AZURE_SIGNTOOL_KEY_VAULT_SECRET%"
```
Same, do we need to quote any of these?

jaimergp commented on constructor/signing.py:
```diff
@@ -0,0 +1,220 @@
+            raise RuntimeError(f"Signature verification failed.\n{proc.stderr}")
+        try:
+            status, status_message = proc.stdout.strip().split("\n")
+            status = int(status)
```
Can this fail if status is empty or None?

marcoesters replied:
I'm not sure how this could be `None` unless `proc.stdout` can somehow be `None`.

If `status` is and empty string, `int(status)` raises a `ValueError`, which is caught further down.

jaimergp commented on tests/test_examples.py:
```diff
@@ -484,6 +483,43 @@ def test_example_signing(tmp_path, request):
+)
+@pytest.mark.parametrize(
+        "auth_method",
+        os.environ.get("AZURE_SIGNTOOL_TEST_AUTH_METHODS", "token,secret").split(","),
```
Why do we need this "parametrized" parametrization? Are you using it locally while debugging? I don't see it in the CI setup 🤔

marcoesters replied:
I wanted to make sure that all authentication methods could potentially be covered. For tokens and secrets, I can just check the environment variables and skip if needed. For managed identities, however, this does not work because there isn't an environment variable to check for (logging into the vault must happen before AzureSignTool is executed).

So, the only "safe" methods to add into the parametrization are `token` and `secret`. However, I still wanted to allow testing with a managed identity.

Does that make sense? I tried to explain this briefly in the docstring.

jaimergp commented on tests/test_examples.py:
```diff
@@ -484,6 +483,43 @@ def test_example_signing(tmp_path, request):
+    """Test signing installers with AzureSignTool.
+
+    There are three ways to authenticate with Azure: tokens, secrets, and managed identities.
+    There is no good sentinel environment for manged identities, so an environment variable
```
```suggestion
    There is no good sentinel environment for managed identities, so an environment variable
```

I think?

jaimergp commented on docs/source/howto.md:
```diff
@@ -38,21 +38,63 @@ On Windows, you can also add extra pages to the installer. This is an advanced o
+
+If neither `AZURE_SIGNTOOL_KEY_VAULT_ACCESSTOKEN` nor `AZURE_SIGNTOOL_KEY_VAULT_SECRET` are set, `constructor` will use a Managed Identity (`-kvm` CLI option).
+:::
+### Signing and notarizing PKG installers
```
```suggestion

### Signing and notarizing PKG installers
```
